File: app/fsharp/tests/Ch08/AsyncIOTests.fs
module Ch08.AsyncIOTests

open Xunit
open Ch08.AsyncIO
open Ch08.MeetingScheduler

// ============================================
// 基本的な Async のテスト
// ============================================

[<Fact>]
let ``delayEffect は副作用をラップする`` () =
    let mutable called = false
    let effect = delayEffect (fun () ->
        called <- true
        42)

    // まだ実行されていない
    Assert.False(called)

    // 実行
    let result = runSync effect
    Assert.True(called)
    Assert.Equal(42, result)

[<Fact>]
let ``pureAsync は純粋な値をラップする`` () =
    let asyncValue = pureAsync 123
    let result = runSync asyncValue
    Assert.Equal(123, result)

[<Fact>]
let ``unitAsync は unit を返す`` () =
    let asyncUnit = unitAsync
    let result = runSync asyncUnit
    Assert.Equal((), result)

// ============================================
// サイコロのテスト
// ============================================

[<Fact>]
let ``castTheDie は1から6の値を返す`` () =
    for _ in 1..100 do
        let result = runSync (castTheDie ())
        Assert.True(result >= 1 && result <= 6)

[<Fact>]
let ``castTheDieTwice は2から12の値を返す`` () =
    for _ in 1..100 do
        let result = runSync (castTheDieTwice ())
        Assert.True(result >= 2 && result <= 12)

// ============================================
// ミーティングスケジューリングのテスト
// ============================================

[<Fact>]
let ``meetingsOverlap は重なりを正しく判定する`` () =
    let meeting1 = { StartHour = 9; EndHour = 10 }
    let meeting2 = { StartHour = 10; EndHour = 11 }
    let meeting3 = { StartHour = 9; EndHour = 11 }

    // 隣接は重ならない
    Assert.False(meetingsOverlap meeting1 meeting2)

    // 重なっている
    Assert.True(meetingsOverlap meeting1 meeting3)
    Assert.True(meetingsOverlap meeting3 meeting2)

[<Fact>]
let ``calendarEntries は予定を返す`` () =
    let aliceEntries = runSync (calendarEntries "Alice")
    Assert.Equal(2, List.length aliceEntries)

    let bobEntries = runSync (calendarEntries "Bob")
    Assert.Equal(2, List.length bobEntries)

    let unknownEntries = runSync (calendarEntries "Unknown")
    Assert.Empty(unknownEntries)

[<Fact>]
let ``scheduledMeetings は2人の予定を結合する`` () =
    let meetings = runSync (scheduledMeetings "Alice" "Bob")
    Assert.Equal(4, List.length meetings)

[<Fact>]
let ``possibleMeetings は空きスロットを計算する`` () =
    let existing = [
        { StartHour = 9; EndHour = 10 }
        { StartHour = 14; EndHour = 15 }
    ]
    let possible = possibleMeetings existing 8 17 1

    // 8-9, 10-11, 11-12, 12-13, 13-14, 15-16, 16-17 が空き
    Assert.True(List.length possible >= 6)
    Assert.True(possible |> List.exists (fun m -> m.StartHour = 8))
    Assert.True(possible |> List.exists (fun m -> m.StartHour = 10))

[<Fact>]
let ``schedule は2人のスケジュールから空きを見つける`` () =
    let slots = runSync (schedule "Alice" "Bob" 1)
    // Alice: 9-10, 14-15
    // Bob: 10-11, 15-16
    // 空き: 8-9, 11-12, 12-13, 13-14
    Assert.True(List.length slots > 0)

// ============================================
// エラーハンドリングのテスト
// ============================================

[<Fact>]
let ``orElse は成功時にメインの値を返す`` () =
    let primary = succeedingAsync 42
    let fallback = succeedingAsync 0
    let result = runSync (primary |> orElse fallback)
    Assert.Equal(42, result)

[<Fact>]
let ``orElse は失敗時にフォールバックを返す`` () =
    let primary = failingAsync "error"
    let fallback = succeedingAsync 99
    let result = runSync (primary |> orElse fallback)
    Assert.Equal(99, result)

// ============================================
// リトライのテスト
// ============================================

[<Fact>]
let ``retry は成功するまでリトライする`` () =
    let mutable attempts = 0
    let action = async {
        attempts <- attempts + 1
        if attempts < 3 then
            failwith "not yet"
        return 42
    }

    let result = runSync (retry 5 action)
    Assert.Equal(42, result)
    Assert.Equal(3, attempts)

[<Fact>]
let ``retryWithDefault は全部失敗したらデフォルト値を返す`` () =
    let action = async {
        failwith "always fails"
        return 42
    }

    let result = runSync (retryWithDefault 3 (-1) action)
    Assert.Equal(-1, result)

// ============================================
// sequence のテスト
// ============================================

[<Fact>]
let ``sequence は Async のリストを List の Async に変換する`` () =
    let asyncList = [
        pureAsync 1
        pureAsync 2
        pureAsync 3
    ]

    let result = runSync (sequence asyncList)
    Assert.Equal<int list>([1; 2; 3], result)

[<Fact>]
let ``scheduledMeetingsMultiple は複数人の予定を取得する`` () =
    let meetings = runSync (scheduledMeetingsMultiple ["Alice"; "Bob"; "Charlie"])
    // Alice: 2, Bob: 2, Charlie: 1
    Assert.Equal(5, List.length meetings)

// ============================================
// 並列実行のテスト
// ============================================

[<Fact>]
let ``parallel2 は2つの Async を並列実行する`` () =
    let async1 = pureAsync 1
    let async2 = pureAsync "hello"
    let result = runSync (parallel2 async1 async2)
    Assert.Equal((1, "hello"), result)

[<Fact>]
let ``parallelAll は Async のリストを並列実行する`` () =
    let asyncList = [
        pureAsync 1
        pureAsync 2
        pureAsync 3
    ]

    let result = runSync (parallelAll asyncList)
    Assert.Equal<int list>([1; 2; 3], result)

// ============================================
// タイムアウトのテスト
// ============================================

[<Fact>]
let ``runWithTimeout は時間内に完了したら Some を返す`` () =
    let fast = pureAsync 42
    let result = runWithTimeout 1000 fast
    Assert.Equal(Some 42, result)

// ============================================
// MeetingScheduler のテスト
// ============================================

[<Fact>]
let ``scheduleAndCreate は最初の空きスロットでミーティングを作成する`` () =
    let result = runSync (scheduleAndCreate "Alice" "Bob" 1)
    Assert.True(result.IsSome)

[<Fact>]
let ``scheduleAndCreate は空きがない場合 None を返す`` () =
    // すべての時間が埋まっている人をシミュレート
    // ここでは既存の実装では発生しないが、ロジックを確認
    let existing = [
        for h in 8..15 -> { StartHour = h; EndHour = h + 1 }
    ]
    let slots = possibleMeetings existing 8 16 1
    Assert.Empty(slots)

File: app/csharp/tests/Ch08/AsyncIOTests.cs
using Ch08;
using LanguageExt;
using static LanguageExt.Prelude;
using Xunit;
using static Ch08.AsyncIO;

namespace GrokkingFP.CSharp.Tests.Ch08;

/// <summary>
/// 第8章: 非同期処理のテスト
/// </summary>
public class AsyncIOTests
{
    // ============================================
    // 基本的な Task のテスト
    // ============================================

    [Fact]
    public async Task PureAsync_ReturnsValue()
    {
        var result = await PureAsync(42);
        Assert.Equal(42, result);
    }

    [Fact]
    public async Task UnitAsync_ReturnsUnit()
    {
        var result = await UnitAsync;
        Assert.Equal(unit, result);
    }

    [Fact]
    public async Task DelayEffect_ExecutesEffect()
    {
        var executed = false;
        var result = await DelayEffect(() =>
        {
            executed = true;
            return 42;
        });
        Assert.True(executed);
        Assert.Equal(42, result);
    }

    // ============================================
    // サイコロのテスト
    // ============================================

    [Fact]
    public void CastTheDieImpure_ReturnsBetween1And6()
    {
        for (var i = 0; i < 100; i++)
        {
            var result = CastTheDieImpure();
            Assert.InRange(result, 1, 6);
        }
    }

    [Fact]
    public async Task CastTheDie_ReturnsBetween1And6()
    {
        for (var i = 0; i < 10; i++)
        {
            var result = await CastTheDie();
            Assert.InRange(result, 1, 6);
        }
    }

    [Fact]
    public async Task CastTheDieTwice_ReturnsBetween2And12()
    {
        for (var i = 0; i < 10; i++)
        {
            var result = await CastTheDieTwice();
            Assert.InRange(result, 2, 12);
        }
    }

    // ============================================
    // ミーティングスケジューリングのテスト
    // ============================================

    [Fact]
    public void MeetingsOverlap_DetectsOverlap()
    {
        var meeting1 = new MeetingTime(9, 10);
        var meeting2 = new MeetingTime(9, 11);
        Assert.True(MeetingsOverlap(meeting1, meeting2));
    }

    [Fact]
    public void MeetingsOverlap_DetectsNoOverlap()
    {
        var meeting1 = new MeetingTime(9, 10);
        var meeting2 = new MeetingTime(10, 11);
        Assert.False(MeetingsOverlap(meeting1, meeting2));
    }

    [Fact]
    public async Task CalendarEntries_ReturnsEntriesForAlice()
    {
        var entries = await CalendarEntries("Alice");
        Assert.Equal(2, entries.Count);
    }

    [Fact]
    public async Task CalendarEntries_ReturnsEmptyForUnknown()
    {
        var entries = await CalendarEntries("Unknown");
        Assert.True(entries.IsEmpty);
    }

    [Fact]
    public async Task ScheduledMeetings_CombinesTwoPersonsEntries()
    {
        var meetings = await ScheduledMeetings("Alice", "Bob");
        Assert.Equal(4, meetings.Count);
    }

    [Fact]
    public void PossibleMeetings_FindsAvailableSlots()
    {
        var existing = Seq(new MeetingTime(9, 10), new MeetingTime(14, 15));
        var possible = PossibleMeetings(existing, 8, 16, 1);
        Assert.Contains(possible, m => m.StartHour == 8);
        Assert.DoesNotContain(possible, m => m.StartHour == 9);
        Assert.Contains(possible, m => m.StartHour == 10);
    }

    [Fact]
    public async Task Schedule_FindsAvailableSlots()
    {
        var slots = await Schedule("Alice", "Bob", 1);
        Assert.True(slots.Count > 0);
    }

    // ============================================
    // OrElse のテスト
    // ============================================

    [Fact]
    public async Task OrElse_ReturnsFirstWhenSucceeds()
    {
        var result = await SucceedingAsync(42).OrElse(SucceedingAsync(100));
        Assert.Equal(42, result);
    }

    [Fact]
    public async Task OrElse_ReturnsFallbackWhenFails()
    {
        var result = await FailingAsync("error").OrElse(SucceedingAsync(100));
        Assert.Equal(100, result);
    }

    // ============================================
    // リトライのテスト
    // ============================================

    [Fact]
    public async Task Retry_ReturnsValueOnSuccess()
    {
        var result = await Retry(3, () => Task.FromResult(42));
        Assert.Equal(42, result);
    }

    [Fact]
    public async Task Retry_RetriesOnFailure()
    {
        var attempts = 0;
        var result = await Retry(3, async () =>
        {
            attempts++;
            if (attempts < 3)
                throw new Exception("fail");
            return 42;
        });
        Assert.Equal(42, result);
        Assert.Equal(3, attempts);
    }

    [Fact]
    public async Task RetryWithDefault_ReturnsDefaultAfterMaxRetries()
    {
        var result = await RetryWithDefault(3, 0, () => Task.FromException<int>(new Exception("fail")));
        Assert.Equal(0, result);
    }

    // ============================================
    // Sequence のテスト
    // ============================================

    [Fact]
    public async Task Sequence_CombinesTasks()
    {
        var tasks = Seq(
            Task.FromResult(1),
            Task.FromResult(2),
            Task.FromResult(3)
        );
        var result = await Sequence(tasks);
        Assert.Equal(Seq(1, 2, 3), result);
    }

    [Fact]
    public async Task ScheduledMeetingsMultiple_CombinesMultipleAttendees()
    {
        var meetings = await ScheduledMeetingsMultiple(Seq("Alice", "Bob", "Charlie"));
        Assert.Equal(5, meetings.Count); // 2 + 2 + 1
    }

    // ============================================
    // 並列実行のテスト
    // ============================================

    [Fact]
    public async Task Parallel2_RunsBothTasks()
    {
        var result = await Parallel2(
            Task.FromResult(1),
            Task.FromResult("hello")
        );
        Assert.Equal((1, "hello"), result);
    }

    [Fact]
    public async Task ParallelAll_RunsAllTasks()
    {
        var tasks = Seq(
            Task.FromResult(1),
            Task.FromResult(2),
            Task.FromResult(3)
        );
        var result = await ParallelAll(tasks);
        Assert.Equal(Seq(1, 2, 3), result);
    }

    // ============================================
    // RunSync のテスト
    // ============================================

    [Fact]
    public void RunSync_ExecutesSynchronously()
    {
        var result = RunSync(Task.FromResult(42));
        Assert.Equal(42, result);
    }

    [Fact]
    public async Task RunWithTimeout_ReturnsValueBeforeTimeout()
    {
        var result = await RunWithTimeout(1000, Task.FromResult(42));
        Assert.True(result.IsSome);
        Assert.Equal(42, result.Match(Some: x => x, None: () => 0));
    }
}

/// <summary>
/// MeetingScheduler のテスト
/// </summary>
public class MeetingSchedulerTests
{
    [Fact]
    public async Task ScheduleAndCreate_ReturnsSlotWhenAvailable()
    {
        var result = await MeetingScheduler.ScheduleAndCreate("Alice", "Charlie", 1);
        Assert.True(result.IsSome);
    }
}
